persist/fs/commitlog/reader.go

```diff
@@ -182,38 +349,49 @@ func (r *reader) readChunk() ([]byte, error) {
 	if _, err := r.chunkReader.Read(buffer); err != nil {
 		return nil, err
 	}
-	return buffer, nil
+
+	// Copy message into pooled byte slice. Reading happens synchronously so its
+	// safe to reuse the same buffer repeatedly, but decoding happens in parallel
+	// so we need to make sure that the byte slice we return won't be mutated
+	// until the caller is done with it.
+	pooled := r.bytesPool.Get(int(size))
+	pooled.IncRef()
+	pooled.AppendAll(buffer)
+	pooled.DecRef()
+	return pooled, nil
 }
 
 func (r *reader) readInfo() (schema.LogInfo, error) {
 	data, err := r.readChunk()
 	if err != nil {
 		return emptyLogInfo, err
 	}
-	r.logDecoder.Reset(data)
-	return r.logDecoder.DecodeLogInfo()
+	data.IncRef()
+	r.logDecoder.Reset(data.Get())
+	logInfo, err := r.logDecoder.DecodeLogInfo()
+	data.DecRef()
+	data.Finalize()
+	return logInfo, err
 }
 
-func (r *reader) readEntry() (schema.LogEntry, error) {
-	data, err := r.readChunk()
-	if err != nil {
-		return emptyLogEntry, err
-	}
-	r.logDecoder.Reset(data)
-	return r.logDecoder.DecodeLogEntry()
+func (r *reader) Close() error {
+	// Shutdown the readLoop goroutine which will shut down the decoderLoops
+	// and close the fd
+	r.cancelFunc()
+
+	return <-r.shutdownChan
 }
 
-func (r *reader) Close() error {
+func (r *reader) close() error {
```

Comment:
Just briefly, is there a test that tests reusing the reader (i.e. Open .. then Close then Open .. Close again)?  Just wanting to make sure we're testing the lifecycle now that we have more complex inner statefulness.

Reply:
I'll add a unit test for this

Reply:
could you add a [leaktest](https://github.com/fortytw2/leaktest) assert to the tests to ensure we're not leaking go-routines

Reply:
Oh dope that looks cool, I'll add it